Only set aria-required to true on mandatory fields

daniel-ac-martin opened the pull request:
This commit changes behaviour to only set aria-required to true, when
the field is set to mandatory, either by setting the 'required' property
or by setting the 'required' validator.

daniel-ac-martin commented:
Apologies, there might be a bug in this PR.
Closing until I have ruled it out, or fixed the problem.

daniel-ac-martin commented:
Apologies again. I was being stupid and running tests against an older version of the code. There is not bug.

Re-opening...

gavboulton commented:
Thanks for this. Can you add some tests please @daniel-ac-martin, especially as it's a breaking change.

daniel-ac-martin commented:
@gavboulton: Will do. I also intend to change this to support the old behaviour using the required property. (I'd previously assumed that this was a bug and that the intention had been to work off the validator rather than have a dedicated property for controlling this.)

daniel-ac-martin commented:
@gavboulton I've updated the code and added a few tests describing the new behaviour.

There are two main changes:
  * In the absence of the required property, the required validator now also sets aria-required to true. Giving forms a free accessibility boost.
  * By default, aria-required is now set to false, making this a breaking change as you pointed out.

gavboulton commented on lib/template-mixins.js:
```diff
@@ -113,7 +113,12 @@ module.exports = function (fields, options) {
+                    ? fields[key].required
+                    : (fields[key].validate && (_.indexOf(fields[key].validate, 'required') !== -1))
+                      ? true : false,
                 pattern: extension.pattern,
```
I think this many if/else ifs deserve their own function.

daniel-ac-martin replied:
Would a "var" do? Or do you want to get rid of one or two of the IF expressions in favour of IF statements?

gavboulton replied:
A var would do and statements instead of expressions.

gavboulton commented on test/spec.index.js:
```diff
@@ -250,6 +250,57 @@ describe('Template Mixins', function () {
+                }));
+            });
+
+            it('the required property takes precedence over the required validator', function () {
```
I'm happy to lose the `required` property in favour of using the validator. Mainly so don't need this test.

daniel-ac-martin replied:
Yeah, but isn't that too tight a coupling to passports-form-wizard? If someone is using template-mixins without form-wizard, having to set the validators list would seem very strange indeed.

daniel-ac-martin replied:
(It also means this is less of a breaking change than it would otherwise be.)

daniel-ac-martin replied:
(Also is there any chance that someone would want different behaviour from template-mixins than from form-wizard?)

gavboulton replied:
True.

What do you mean about 'different behaviour from the template-mixins'?

daniel-ac-martin replied:
If they wanted aria-required set to false but the validator to function, or if they wanted aria-required set to true but not enforce this in the validators. (Perhaps they are using some custom logic to validate the step using a custom controller.)

gavboulton replied:
Maybe we just stick with the `required` property then. This would enable the most flexibility.

gavboulton commented on test/spec.index.js:
```diff
@@ -250,6 +250,57 @@ describe('Template Mixins', function () {
                 }));
             });
 
+            it('by default, assumes the field isn\'t required', function () {
```
We should be defaulting to true here.

daniel-ac-martin replied:
That's the old behaviour. But it seems strange to me to assume that a field is required.

If we keep defaulting to true then this PR is pointless (without extra logic). i.e. We test for the presence of the 'required' validator and set aria-required to true only when it is present. If we default to true there is no point in changing anything.

UNLESS, we default to true, and only set to false when fields[key].validate is set but does not contain the required validator. The would make this even less of a breaking change, but might be quite weird behaviour.

gavboulton replied:
By default, form inputs across government are mandatory.

daniel-ac-martin replied:
In that case, should form-wizard not default to applying the 'required' validator? i.e.That the user of the library must actively state that the field is optional?

It seems to me that no matter how you look at it, there is a mismatch between how the two libraries work; template-mixins defaults to required, form-wizard / form-controller defaults to optional.

Should these libraries be attempting to enforce these government standards? Is that not the job of the team building the individual form?

gavboulton replied:
There's definitely a mismatch.How would you override a 'required' validator? Following an approach that requires the minimum configuration makes the most sense to me.

daniel-ac-martin replied:
Well quite. It's ugly to change the form-wizard functionality and create an 'anti-validator' though it could be done.

My personal opinion is that this PR should be merged and we should default to fields being optional from a technological standpoint, and put the onus back on development teams to only ask for the information they need. (Which IMHO is where it belongs. The libraries should reflect the nature of the web and not the nature of current government policies. That should be enforced at another level.)

BTW, presumably all projects using these libraries suffer from the bug reported on BRP and this PR should fix it for them. Which is pretty nice.

There's also the issue that enforcing government policy in the libraries themselves makes them far less useful outside of government. - I don't see any reason why a lot of this stuff shouldn't be used out in the wider world!

gavboulton replied:
Cool. This would bring the the template-mixins in line with form-wizard. I don't foresee any implementation problems so I'll merge and publish as a major version.

gavboulton commented:
hmpo-template-mixins@4.0.0 published.

daniel-ac-martin commented:
@gavboulton Thanks.
